README.md
btc-script
==========

JavaScript component for Bitcoin script.



Install
-------

### Node.js/Browserify

    npm install --save btc-script




Usage
-----

(TODO)



Credits
-------

Most of the code from Stefan Thomas in https://github.com/bitcoinjs/bitcoinjs-lib



License
-------

(MIT License)




CHANGELOG.md
0.1.11 / 2014-11-20
-------------------
* Allow to set explicit sorting for public keys in multisig (#13)

0.1.10 / 2014-05-13
------------------
Bugfix - clone script in .toASM()

0.1.9 / 2014-04-10
------------------
Bugfix - detect zero byte pubkeys in a nonstandard multisig

0.1.8 / 2014-04-10
-----------------
* ASM support

0.1.7 / 2014-04-09
-----------------
* Multisig Escrows Addresses Support

0.1.4 / 2014-04-05
-----------------
* Support OPRETURN

0.1.3 / 2014-02-19
------------------
* Bugfix - Add script validation rules; exclude scripts over 10,000 bytes

0.1.2 / 2014-0218
-----------------
* Bugfix - better validation for standard output script types
* Bugfix - optional field to indicate to not parse BIP34 coinbase input scripts

0.1.1 / 2014-03-13
------------------
* Made script type setting an explicit parameter passed to `createOutputScript()` rather than guessing based on version
* removed method `simpleInHash()`. Closes #1
* removed method `simpleInPubkeyHash()`
* fixed correct output type in `getOutType()` Closes #6
* added `defaultNetworkType`, implicitly sets `Address.defaultNetworkType`
* replaced `convert-hex` for `binstring`
* (straight to 0.1.1, bug in NPM: https://github.com/npm/npm/issues/4653)


0.0.2 / 2014-02-04
------------------
* added fromChunks method to produce a script object from chunks array

0.0.1 / 2014-01-12
------------------
* initial release




test/script.test.js
var Script = require('../')
var binConv = require('binstring')
var _ = require('underscore')
var Address = require('btc-address')
var sha256ripe160 = require('crypto-hashing').sha256ripe160

require('terst')

describe('Script', function() {
  describe(' - fromChunks()', function() {
    it(' > generates a Script object from chunks', function() {
      var script = new Script('76a914652c453e3f8768d6d6e1f2985cb8939db91a4e0588ac')
      var tmp = script.clone()
      var scriptFromChunks = Script.fromChunks(tmp.chunks)

      // check buffers are the same
      T(script.buffer.length === scriptFromChunks.buffer.length)
      for (var i in scriptFromChunks.buffer) {
        T(script.buffer[i] === scriptFromChunks.buffer[i])
      }

      // check chunks are the same
      T(script.chunks.length === scriptFromChunks.chunks.length)
      for (var i in scriptFromChunks.chunks) {
        // chunks contain nested arrays
        if (!Array.isArray(scriptFromChunks.chunks[i])) {
          T(script.chunks[i] === scriptFromChunks.chunks[i])
        } else {
          T(script.chunks[i].length === scriptFromChunks.chunks[i].length)
          var scriptFromChunkArray = scriptFromChunks.chunks[i]
          var scriptArray = script.chunks[i]
          for (var y in scriptFromChunkArray[y]) {
            T(scriptArray[i] === scriptFromChunkArray[i])
          }
        }
      }
    })
  })

  describe(' - Script.createOutputScript()', function() {
    it(' > Supports pubkeyhash addresses', function() {
      // hash160 : 0000000000000000000000000000000000000000
      var pubkeyScript = Script.createOutputScript('1111111111111111111114oLvT2')
      EQ(pubkeyScript.chunks[2].length, 20)
      for (var i in pubkeyScript.chunks[2]) {
        EQ(pubkeyScript.chunks[2][i], 0)
      }
      EQ(pubkeyScript.getOutType(), 'pubkeyhash')
    })

    it(' > Supports scripthash addresses', function() {
      // hash160 : 0000000000000000000000000000000000000000
      var scripthashScript = Script.createOutputScript('31h1vYVSYuKP6AhS86fbRdMw9XHieotbST')
      EQ(scripthashScript.chunks[1].length, 20)
      for (var i in scripthashScript.chunks[1]) {
        EQ(scripthashScript.chunks[1][i], 0)
      }
      EQ(scripthashScript.getOutType(), 'scripthash')
    })
  })

  describe('- Script.getOutType()', function() {
    it(' > Supports Pubkeyhash', function() {
      // https://helloblock.io/transactions/dc55d9c6ec03ceccf0db43d29e7d626a8b107f41066e3917f30398bb01dda2b5
      var outHex = '76a9140c79f15f08b0f94a2cba8de036272dee3ecb096188ac'
      var pubkeyhashScript = new Script(outHex)
      EQ(pubkeyhashScript.getOutType(), 'pubkeyhash')
    })

    it(' > Supports Pubkey', function() {
      // https://helloblock.io/transactions/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098
      var outHex = '410496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858eeac'
      var pubkeyScript = new Script(outHex)
      EQ(pubkeyScript.getOutType(), 'pubkey')
    })

    it(' > Supports Scripthash', function() {
      // https://helloblock.io/transactions/64ebe48e0c9fb8da32160809848a464d303b1e24026d2359597889c3f46f2e21
      var outHex = 'a914f9659d50ec62945a9c8a6c07f117ef63a4f2fcc487'
      var scripthashScript = new Script(outHex)
      EQ(scripthashScript.getOutType(), 'scripthash')
    })

    it(' > Supports Multisig', function() {
      // https://helloblock.io/transactions/09dd94f2c85262173da87a745a459007bb1eed6eeb6bfa238a0cd91a16cf7790
      var outHex = '5121032487c2a32f7c8d57d2a93906a6457afd00697925b0e6e145d89af6d3bca330162102308673d16987eaa010e540901cc6fe3695e758c19f46ce604e174dac315e685a52ae'
      var multisigScript = new Script(outHex)
      EQ(multisigScript.getOutType(), 'multisig')
    })

    it(' > Supports invalid Multisig', function() {
      // https://helloblock.io/transactions/dfa8ff97f33cb83dbaa22ed3a99883218d6afd681d486b374496d145b39a63b7
      // asm: "0 0 0 OP_CHECKMULTISIG"
      var outHex = '000000ae'
      var multisigScript = new Script(outHex)
      EQ(multisigScript.getOutType(), 'nonstandard')
    })

    it(' Supports invalid Multisig', function() {
      // https://helloblock.io/transactions/4142ee4877eb116abf955a7ec6ef2dc38133b793df762b76d75e3d7d4d8badc9
      var outHex = '52210352f2e34760514099f90b03aab91239466924c3b06047d3cf0e011f26ef96ceb7210207cf5fb65d6923d5d41db21ceac9567a0fc3eb92c6137f274018381ced7b65680053ae'
      var multisigScript = new Script(outHex)
      EQ(multisigScript.getOutType(), 'nonstandard')
      // asm: 2 0352f2e34760514099f90b03aab91239466924c3b06047d3cf0e011f26ef96ceb7 0207cf5fb65d6923d5d41db21ceac9567a0fc3eb92c6137f274018381ced7b6568 0 3 OP_CHECKMULTISIG

    })

    it(' > Supports null_data (OP_RETURN)', function() {
      var outHex = '6a4c105768792068656c6c6f2074686572652e'
      var nulldataScript = new Script(outHex)
      EQ(nulldataScript.getOutType(), 'nulldata')
    })

    it(' > Supports null_data (OP_RETURN)', function() {
      // https://helloblock.io/transactions/5e9be7fb36ee49ce84bee4c8ef38ad0efc0608b78dae1c2c99075297ef527890
      // op_return
      var outHex = '6a2606deadbeef03f895a2ad89fb6d696497af486cb7c644a27aa568c7a18dd06113401115185474'
      var opreturnScript = new Script(outHex)
      EQ(opreturnScript.getOutType(), 'nulldata')
    })

    it(' > nonstandard (zero byte as hash160 in pubkeyhash)', function() {
      // http://testnet.helloblock.io/transactions/a347b4ef02173b74deb096921d8306ff7c379c254e9febaa040024b220a348ed
      var nonstandardpubkeyhash = new Script('76a90088ac')
      EQ(nonstandardpubkeyhash.getOutType(), 'nonstandard')
    })

    it(' > nonstandard (zero byte as hash160 in scripthash)', function() {
      var nonstandardpubkeyhash = new Script('a90087')
      EQ(nonstandardpubkeyhash.getOutType(), 'nonstandard')
    })
  })

  describe('- Script.toAddress()', function() {
    it(' > returns address for pubkeyScript', function() {
      // https://helloblock.io/transactions/0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098
      var outHex = '410496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858eeac'
      var pubkeyScript = new Script(outHex)
      EQ(pubkeyScript.getOutType(), 'pubkey')

      var addr = pubkeyScript.toAddress()
      EQ(binConv(addr.hash, { in : 'bytes',
        out: 'hex'
      }), '119b098e2e980a229e139a9ed01a469e518e6f26')
    })
  })

  describe(' - Script.toAddresses()', function() {
    it(' > returns addresses for multisig', function() {
      // https://helloblock.io/transactions/09dd94f2c85262173da87a745a459007bb1eed6eeb6bfa238a0cd91a16cf7790
      var outHex = '5121032487c2a32f7c8d57d2a93906a6457afd00697925b0e6e145d89af6d3bca330162102308673d16987eaa010e540901cc6fe3695e758c19f46ce604e174dac315e685a52ae'
      var multisigScript = new Script(outHex)
      EQ(multisigScript.getOutType(), 'multisig')
      var addrs = multisigScript.toAddresses()
      EQ(addrs.length, 2);
      EQ(addrs[0].toString(), '17rExRiMaJGgHPVuYjHL21mEhfzbRPRkui')
      EQ(addrs[1].toString(), '1GUx7qydsHMWaMtF728HfSX1EQpQrZccdb')

      // https://test.helloblock.io/transactions/f823bd777b9a3e4a7dda9523ad90785e5bc68710c173788f21e2fe2df3cfea8f
      var outHex = '52210266816de738c62ad789119fdb13131faa13f588359484ca61d0515cdcc7648ecd21025fe4a325d96f109529734af5de80b961274de5720c30646c398202e5d555adca52ae'
      var multisigScript = new Script(outHex)
      EQ(multisigScript.getOutType(), 'multisig')
      var addrs = multisigScript.toAddresses('testnet')
      EQ(addrs.length, 2);
      EQ(addrs[0].toString(), 'n2QXt6v6F9LTLCTjD6ZBHynWcx7Fo8991s')
      EQ(addrs[1].toString(), 'moiaQ6szVvA1ccC25eo26jnHxr8R1Jt5Ze')
    })
    it('> returns addresses for pubkeyhash', function() {
      var outHex = '76a9140c79f15f08b0f94a2cba8de036272dee3ecb096188ac'
      var pubkeyhashScript = new Script(outHex)
      EQ(pubkeyhashScript.toAddresses()[0].toString(), '128yC3L6zguTqfKTA1z18EuGMoV5tkHkR6')
    })
  })

  describe(' - Setting defaultNetwork', function() {
    it(' > defaults to mainnet and supports testnet', function() {
      // hash160: 0000000000000000000000000000000000000000
      Script.defaultNetwork = 'testnet'
      var pubkeyhashScript = Script.createOutputScript('mfWxJ45yp2SFn7UciZyNpvDKrzbhyfKrY8')
      EQ(pubkeyhashScript.chunks[2].length, 20)
      for (var i in pubkeyhashScript.chunks[2]) {
        EQ(pubkeyhashScript.chunks[2][i], 0)
      }
      EQ(pubkeyhashScript.getOutType(), 'pubkeyhash')
    })
  })

  describe(' - Script(hex, isCoinbase)', function() {
    it(' > optional second field to indicate coinbase and does not interpret coinbase script', function() {
      // reference:
      // https://github.com/cryptocoinjs/btc-script/issues/8
      // BIP34: https: //github.com/bitcoin/bips/blob/master/bip-0034.mediawiki
      // Stackoverflow: http: //bitcoin.stackexchange.com/questions/20721/what-is-the-format-of-coinbase-transaction

      // optional second field
      var coinbaseScript = new Script('03d891000450a8eb0b4ed20100', true)
      T(_.isEqual(coinbaseScript.buffer, coinbaseScript.buffer))
    })

    it(' > getBlockHeight() is able to parse blockheight', function() {
      // http://blockexplorer.com/testnet/tx/eaff2bbc220caa80f1ee52b2ed3fb586b26e253d2dee039fd6c54ec8493aa1d8
      var coinbaseScript = new Script('03d891000450a8eb0b4ed20100', true)
      EQ(coinbaseScript.getBlockHeight(), 37336)
    })
  })

  describe(' - toASM', function() {
    it(' > supports standard script type', function() {
      var scripts = [{
        hex: 'a914f9659d50ec62945a9c8a6c07f117ef63a4f2fcc487',
        asm: 'OP_HASH160 f9659d50ec62945a9c8a6c07f117ef63a4f2fcc4 OP_EQUAL'
      }, {
        hex: '410496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858eeac',
        asm: '0496b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52da7589379515d4e0a604f8141781e62294721166bf621e73a82cbf2342c858ee OP_CHECKSIG'
      }, {
        hex: '5121032487c2a32f7c8d57d2a93906a6457afd00697925b0e6e145d89af6d3bca330162102308673d16987eaa010e540901cc6fe3695e758c19f46ce604e174dac315e685a52ae',
        asm: 'OP_1 032487c2a32f7c8d57d2a93906a6457afd00697925b0e6e145d89af6d3bca33016 02308673d16987eaa010e540901cc6fe3695e758c19f46ce604e174dac315e685a OP_2 OP_CHECKMULTISIG'
      }, {
        hex: '76a9140c79f15f08b0f94a2cba8de036272dee3ecb096188ac',
        asm: 'OP_DUP OP_HASH160 0c79f15f08b0f94a2cba8de036272dee3ecb0961 OP_EQUALVERIFY OP_CHECKSIG'
      }, {
        hex: '48304502203cb2b3dfeddaa2790d5ee4bf91fe42c429c583e1189905cfbfe3c079d514ddc0022100d002b19049286e0201e847774bb99e7e18486e66b6b75dd0f4d6e494597b8c660121032487c2a32f7c8d57d2a93906a6457afd00697925b0e6e145d89af6d3bca33016',
        asm: '304502203cb2b3dfeddaa2790d5ee4bf91fe42c429c583e1189905cfbfe3c079d514ddc0022100d002b19049286e0201e847774bb99e7e18486e66b6b75dd0f4d6e494597b8c6601 032487c2a32f7c8d57d2a93906a6457afd00697925b0e6e145d89af6d3bca33016'
      }]

      for (var i = 0; i < scripts.length; i++) {
        var s = scripts[i]
        var script = new Script(s.hex)
        EQ(script.toASM(), s.asm)
      }
    })
  })
  describe(' - Script.createMultiSigOutputScript', function() {
    // sorted in the way createMultiSigOutputScript sorts them internally
    var pubkeys = [ "026b0964dd480712cced3fa2938afb1c25bf2ff893a7efab611de53bb582e422f9",
                    "03822a3e30f86e30a0c7de273cca0201f0b9e8e629834b172cf1e8ef19d4695f31",
                    "03299a2ecb85d4d435cbe0f033780bd5c57cc216afed291f59e89a14f9845f8f0e"
                  ].map(function(pubkey){
                    return binConv(pubkey, { in: 'hex', out: 'bytes' })
                  })
    var unsorted_pubkeys = [ pubkeys[2], pubkeys[0], pubkeys[1] ]
    it(' > Creates multisig scripts', function() {
      var script = Script.createMultiSigOutputScript(2, pubkeys)
      EQ(toAddress(script), '34qjQhsQbXHhD9fZwY8YmFSLiiQcNJeLwQ')
    })
    it(' > Has implicit sorting', function() {
      var script = Script.createMultiSigOutputScript(2, unsorted_pubkeys)
      EQ(toAddress(script), '34qjQhsQbXHhD9fZwY8YmFSLiiQcNJeLwQ')
    })
    it(' > Allows to specify explicit sorting', function() {
      var script = Script.createMultiSigOutputScript(2, unsorted_pubkeys, true)
      EQ(toAddress(script), '3JGpyvuKUoK6aK38zZey6anGdY6DkQnv3s')
      var script = Script.createMultiSigOutputScript(2, pubkeys, true)
      EQ(toAddress(script), '34qjQhsQbXHhD9fZwY8YmFSLiiQcNJeLwQ')
    })
    it(' > Does not mutate the public keys', function() {
      EQ(new Buffer(unsorted_pubkeys[0]).toString('hex'), new Buffer(pubkeys[2]).toString('hex'))
      EQ(new Buffer(unsorted_pubkeys[1]).toString('hex'), new Buffer(pubkeys[0]).toString('hex'))
      EQ(new Buffer(unsorted_pubkeys[2]).toString('hex'), new Buffer(pubkeys[1]).toString('hex'))
    })

    function toAddress(script) {
      return '' + new Address(sha256ripe160(script.buffer, { out: 'bytes' }), 'scripthash', 'mainnet');
    }
  })
})


lib/script.js
var Opcode = require('btc-opcode');
var conv = require('binstring');
var Address = require('btc-address');
var cryptoHash = require('crypto-hashing')
var BigInteger = require('bigi')

var _defaultNetwork = 'mainnet';
Address.defaultNetwork = _defaultNetwork;

module.exports = Script;
Object.defineProperty(Script, 'defaultNetwork', {
  set: function(val) {
    _defaultNetwork = val;
    Address.defaultNetwork = val;
  },
  get: function() {
    return _defaultNetwork;
  }
})

function Script(data, isCoinbase) {
  if (!data) {
    this.buffer = [];
  } else if ("string" == typeof data) {
    this.buffer = conv(data, { in : 'hex',
      out: 'bytes'
    });
  } else if (Array.isArray(data)) {
    this.buffer = data;
  } else if (data instanceof Script) {
    this.buffer = data.buffer;
  } else {
    throw new Error("Invalid script");
  }

  if (isCoinbase) {
    /*
      This only applies to coinbase scripts
      For more information:
      BIP34:
      https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki

      Stackoverflow:
      http://bitcoin.stackexchange.com/questions/20721/what-is-the-format-of-coinbase-transaction
    */

    // if coinbase do not parse
    this.isCoinbase = true
    this.chunks = this.buffer;
    return
  }

  if (!Script.isValid(this.buffer)) {
    this.chunks = this.buffer
    return
  }

  this.parse();
}

Script.isValid = function(buffer) {
  if (Array.isArray(buffer)) {
    // todo: add more validation rules
    if (buffer.length <= 10000) {
      return true
    }
  }
  return false
}

Script.fromPubKey = function(str) {
  var script = new Script();
  var s = str.split(" ");
  for (var i in s) {
    if (Opcode.map.hasOwnProperty(s[i])) {
      script.writeOp(Opcode.map[s[i]]);
    } else {
      script.writeBytes(conv(s[i], { in : 'hex',
        out: 'bytes'
      }));
    }
  }
  return script;
};

Script.fromScriptSig = function(str) {
  var script = new Script();
  var s = str.split(" ");
  for (var i in s) {
    if (Opcode.map.hasOwnProperty(s[i])) {
      script.writeOp(Opcode.map[s[i]]);
    } else {
      script.writeBytes(conv(s[i], { in : 'hex',
        out: 'bytes'
      }));
    }
  }
  return script;
};

Script.fromChunks = function(chunks) {
  var script = new Script();

  for (var i = 0, l = chunks.length; i < l; i++) {
    var data = chunks[i];
    script.writeBytes(data)
  }

  return script;
};

/**
 * Update the parsed script representation.
 *
 * Each Script object stores the script in two formats. First as a raw byte
 * array and second as an array of "chunks", such as opcodes and pieces of
 * data.
 *
 * This method updates the chunks cache. Normally this is called by the
 * constructor and you don't need to worry about it. However, if you change
 * the script buffer manually, you should update the chunks using this method.
 */
Script.prototype.parse = function() {
  var self = this;

  this.chunks = [];

  // Cursor
  var i = 0;

  // Read n bytes and store result as a chunk
  function readChunk(n) {
    self.chunks.push(self.buffer.slice(i, i + n));
    i += n;
  };

  while (i < this.buffer.length && i >= 0) {
    var opcode = this.buffer[i++];
    if (opcode >= 0xF0) {
      // Two byte opcode
      opcode = (opcode << 8) | this.buffer[i++];
    }

    var len;
    if (opcode > 0 && opcode < Opcode.map.OP_PUSHDATA1) {
      // Read some bytes of data, opcode value is the length of data
      readChunk(opcode);
    } else if (opcode == Opcode.map.OP_PUSHDATA1) {
      len = this.buffer[i++];
      readChunk(len);
    } else if (opcode == Opcode.map.OP_PUSHDATA2) {
      len = (this.buffer[i++] << 8) | this.buffer[i++];
      readChunk(len);
    } else if (opcode == Opcode.map.OP_PUSHDATA4) {
      len = (this.buffer[i++] << 24) |
        (this.buffer[i++] << 16) |
        (this.buffer[i++] << 8) |
        this.buffer[i++];
      readChunk(len);
    } else {
      this.chunks.push(opcode);
    }
  }
};

/**
 * Compare the script to known templates of scriptPubKey.
 *
 * This method will compare the script to a small number of standard script
 * templates and return a string naming the detected type.
 *
 * Currently supported are:
 * Pubkeyhash
 *   Paying to a Bitcoin address which is the hash of a pubkey.
 *   OP_DUP OP_HASH160 [pubKeyHash] OP_EQUALVERIFY OP_CHECKSIG
 *   Example:
 *
 * Pubkey:
 *   Paying to a public key directly.
 *   [pubKey] OP_CHECKSIG
 *   Example: txid 0e3e2357e806b6cdb1f70b54c3a3a17b6714ee1f0e68bebb44a74b1efd512098
 *
 * Scripthash (P2SH)
 *    Paying to an address which is the hash of a script
 *    OP_HASH160 [Scripthash] OP_EQUAL
 *    Example:
 *
 * Multisig
 *    Paying to multiple pubkeys and require a number of the signatures
 *    m [pubkey] [pubkey] [pubkey] n OP_CHECKMULTISIG
 *    Example:
 *
 * Strange:
 *   Any other script (no template matched).
 */

// Below is the current standard set of out types
/* const char* GetTxnOutputType(txnouttype t)
{
    switch (t)
    {
    case TX_NONSTANDARD: return "nonstandard";
    case TX_PUBKEY: return "pubkey";
    case TX_PUBKEYHASH: return "pubkeyhash";
    case TX_SCRIPTHASH: return "scripthash";
    case TX_MULTISIG: return "multisig";
    case TX_NULL_DATA: return "nulldata";
    }
    return NULL;
}*/

// https://github.com/bitcoin/bitcoin/blob/19e5b9d2dfcac4efadba636745485d9660fb1abe/src/script.cpp#L75

// supporting tx_null_data https://github.com/bitcoin/bitcoin/pull/3128
// https://helloblock.io/mainnet/transactions/ebc9fa1196a59e192352d76c0f6e73167046b9d37b8302b6bb6968dfd279b767
Script.prototype.getOutType = function() {
  if (this.chunks.length == 5 &&
    this.chunks[0] == Opcode.map.OP_DUP &&
    this.chunks[1] == Opcode.map.OP_HASH160 &&
    Array.isArray(this.chunks[2]) &&
    this.chunks[2].length === 20 &&
    this.chunks[3] == Opcode.map.OP_EQUALVERIFY &&
    this.chunks[4] == Opcode.map.OP_CHECKSIG) {
    // Transfer to Bitcoin address
    return 'pubkeyhash';
  } else if (this.chunks.length === 2 &&
    Array.isArray(this.chunks[0]) &&
    this.chunks[1] === Opcode.map.OP_CHECKSIG) {
    // [pubkey] OP_CHECKSIG
    return 'pubkey';
  } else if (this.chunks[this.chunks.length - 1] == Opcode.map.OP_EQUAL &&
    this.chunks[0] == Opcode.map.OP_HASH160 &&
    Array.isArray(this.chunks[1]) &&
    this.chunks[1].length === 20 &&
    this.chunks.length == 3) {
    // Transfer to M-OF-N
    return 'scripthash';
  } else if (this.chunks.length > 3 &&
    // m is a smallint
    isSmallIntOp(this.chunks[0]) &&
    // n is a smallint
    isSmallIntOp(this.chunks[this.chunks.length - 2]) &&
    // n greater or equal to m
    this.chunks[0] <= this.chunks[this.chunks.length - 2] &&
    // n cannot be 0
    this.chunks[this.chunks.length - 2] !== Opcode.map.OP_0 &&
    // n is the size of chunk length minus 3 (m, n, OP_CHECKMULTISIG)
    this.chunks.length - 3 === this.chunks[this.chunks.length - 2] - Opcode.map.OP_RESERVED &&
    // the middle chunks are all pubkeys
    arePubkeys(this.chunks.slice(1, this.chunks.length - 2)) &&
    // last chunk is OP_CHECKMULTISIG
    this.chunks[this.chunks.length - 1] == Opcode.map.OP_CHECKMULTISIG) {
    return 'multisig'
  } else if (this.chunks[0] === Opcode.map.OP_RETURN) {
    return 'nulldata'
  } else {
    return 'nonstandard';
  }
}

function arePubkeys(pubkeys) {
  for (var i = 0; i < pubkeys.length; i++) {
    if (!Array.isArray(pubkeys[i])) {
      return false
    }
  }
  return true
}

function isSmallIntOp(opcode) {
  return ((opcode == Opcode.map.OP_0) ||
    ((opcode >= Opcode.map.OP_1) && (opcode <= Opcode.map.OP_16)));
};

/**
 * Returns the address corresponding to this output in hash160 form.
 * Assumes strange scripts are P2SH
 */
/*Script.prototype.toScriptHash = function () {
    var outType = this.getOutType();

    return outType == 'Pubkey'       ? this.chunks[2]
         : outType == 'P2SH'         ? util.sha256ripe160(this.buffer)
         :                             util.sha256ripe160(this.buffer)
};*/

Script.prototype.toAddress = function(network) {
  var outType = this.getOutType();
  if (outType == 'pubkeyhash') {
    return new Address(this.chunks[2], 'pubkeyhash', network || Script.defaultNetwork)
  } else if (outType == 'pubkey') {
    // convert pubkey into a pubkeyhash and do address
    return new Address(cryptoHash.sha256ripe160(this.chunks[0], {
        out: 'bytes'
      }),
      'pubkeyhash', network || Script.defaultNetwork)
  } else if (outType == 'scripthash') {
    return new Address(this.chunks[1], 'scripthash', network || Script.defaultNetwork)
  } else {
    return false
  }
}

/*
  returns an array of addresses
*/
Script.prototype.toAddresses = function(network) {
  var self = this;
  var outType = this.getOutType();
  var addresses = [];
  if (outType === 'multisig') {
    for (var i = 1, last = self.chunks.length - 2; i < last; i++) {
      addresses.push(new Address(cryptoHash.sha256ripe160(self.chunks[i], {
          out: 'bytes'
        }),
        'pubkeyhash', network || Script.defaultNetwork))
    }
    return addresses
  } else {
    var address = self.toAddress()
    if (address) {
      return [address]
    } else {
      return false
    }
  }
}

/**
 * Compare the script to known templates of scriptSig.
 *
 * This method will compare the script to a small number of standard script
 * templates and return a string naming the detected type.
 *
 * WARNING: Use this method with caution. It merely represents a heuristic
 * based on common transaction formats. A non-standard transaction could
 * very easily match one of these templates by accident.
 *
 * Currently supported are:
 * Address:
 *   Paying to a Bitcoin address which is the hash of a pubkey.
 *   [sig] [pubKey]
 *
 * Pubkey:
 *   Paying to a public key directly.
 *   [sig]
 *
 * Multisig:
 *   Paying to M-of-N public keys.
 *
 * Strange:
 *   Any other script (no template matched).
 */
Script.prototype.getInType = function() {
  if (this.chunks.length == 1 &&
    Array.isArray(this.chunks[0])) {
    // Direct IP to IP transactions only have the signature in their scriptSig.
    // TODO: We could also check that the length of the data is correct.
    return 'pubkey';
  } else if (this.chunks.length == 2 &&
    Array.isArray(this.chunks[0]) &&
    Array.isArray(this.chunks[1])) {
    return 'pubkeyhash';
  } else if (this.chunks[0] == Opcode.map.OP_0 &&
    this.chunks.slice(1).reduce(function(t, chunk, i) {
      return t && Array.isArray(chunk) && (chunk[0] == 48 || i == this.chunks.length - 1);
    }, true)) {
    return 'multisig';
  } else {
    return 'nonstandard';
  }
};

/**
 * Returns the affected public key for this input.
 *
 * This currently only works with payToPubKeyHash transactions. It will also
 * work in the future for standard payToScriptHash transactions that use a
 * single public key.
 *
 * However for multi-key and other complex transactions, this will only return
 * one of the keys or raise an error. Therefore, it is recommended for indexing
 * purposes to use Script#simpleInHash or Script#simpleOutHash instead.
 *
 * @deprecated
 */
Script.prototype.simpleInPubKey = function() {
  switch (this.getInType()) {
    case 'pubkeyhash':
      return this.chunks[1];
    case 'pubkey':
      // TODO: Theoretically, we could recover the pubkey from the sig here.
      //       See https://bitcointalk.org/?topic=6430.0
      throw new Error("Script does not contain pubkey.");
    default:
      throw new Error("Encountered non-standard scriptSig");
  }
};

/**
 * Add an op code to the script.
 */
Script.prototype.writeOp = function(opcode) {
  this.buffer.push(opcode);
  this.chunks.push(opcode);
};

/**
 * Add a data chunk to the script.
 */
Script.prototype.writeBytes = function(data) {
  // if it is not array then it is opcode
  if (!Array.isArray(data)) {
    this.buffer.push(data);
    this.chunks.push(data);
    return
  }

  if (data.length < Opcode.map.OP_PUSHDATA1) {
    this.buffer.push(data.length);
  } else if (data.length <= 0xff) {
    this.buffer.push(Opcode.map.OP_PUSHDATA1);
    this.buffer.push(data.length);
  } else if (data.length <= 0xffff) {
    this.buffer.push(Opcode.map.OP_PUSHDATA2);
    this.buffer.push(data.length & 0xff);
    this.buffer.push((data.length >>> 8) & 0xff);
  } else {
    this.buffer.push(Opcode.map.OP_PUSHDATA4);
    this.buffer.push(data.length & 0xff);
    this.buffer.push((data.length >>> 8) & 0xff);
    this.buffer.push((data.length >>> 16) & 0xff);
    this.buffer.push((data.length >>> 24) & 0xff);
  }
  this.buffer = this.buffer.concat(data);
  this.chunks.push(data);
};

/**
 * Create an output for an address
 */
Script.createOutputScript = function(address, network) {
  var script = new Script();
  if ('string' === typeof address) {
    // null as we have a string address
    var address = new Address(address, null, network || Script.defaultNetwork);
  }

  var type = address.getType(network || Script.defaultNetwork);

  // Standard pay-to-pubkey-hash
  if (type === 'pubkeyhash') {
    script.writeOp(Opcode.map.OP_DUP);
    script.writeOp(Opcode.map.OP_HASH160);
    script.writeBytes(address.hash);
    script.writeOp(Opcode.map.OP_EQUALVERIFY);
    script.writeOp(Opcode.map.OP_CHECKSIG);
  }

  // Standard pay-to-script-hash
  else if (type === 'scripthash') {
    script.writeOp(Opcode.map.OP_HASH160);
    script.writeBytes(address.hash);
    script.writeOp(Opcode.map.OP_EQUAL);
  } else {
    // address can only be a pubkeyhash address or a scripthash address
    return false
  }

  return script;
};

/**
 * Extract pubkeys from a multisig script
 */

Script.prototype.extractPubkeys = function() {
  return this.chunks.filter(function(chunk) {
    return (chunk[0] == 4 && chunk.length == 65 || chunk[0] < 4 && chunk.length == 33)
  });
}

/**
 * Create an m-of-n output script
 */
Script.createMultiSigOutputScript = function(m, pubkeys, explicitSorting) {
  var script = new Script();

  explicitSorting || (pubkeys = pubkeys.slice().sort());

  script.writeOp(Opcode.map.OP_1 + m - 1);

  for (var i = 0; i < pubkeys.length; ++i) {
    script.writeBytes(pubkeys[i]);
  }

  script.writeOp(Opcode.map.OP_1 + pubkeys.length - 1);

  script.writeOp(Opcode.map.OP_CHECKMULTISIG);

  return script;
};

/**
 * Create a standard payToPubKeyHash input.
 */
Script.createInputScript = function(signature, pubKey) {
  var script = new Script();
  script.writeBytes(signature);
  script.writeBytes(pubKey);
  return script;
};

/**
 * Create a multisig input
 */
Script.createMultiSigInputScript = function(signatures, script) {
  script = new Script(script);
  var k = script.chunks[0][0];
  if (signatures.length < k) return false; //Not enough sigs
  var inScript = new Script();
  inScript.writeOp(Opcode.map.OP_0);
  signatures.map(function(sig) {
    inScript.writeBytes(sig)
  });
  inScript.writeBytes(script.buffer);
  return inScript;
}

Script.prototype.clone = function() {
  return new Script(this.buffer);
};

Script.prototype.getBlockHeight = function() {
  if (this.isCoinbase) {
    // This will never reach PUSHDATA1 so read first byte as length byte and read the rest
    // first byte should stay as 03 for the next 300 years
    var len = this.buffer[0]
    return parseInt(BigInteger.fromByteArrayUnsigned(this.buffer.slice(1, 1 + len).reverse()).toString(10))
  }
  return false
}

Script.prototype.toASM = function(truncate, maxEl) {
  var scriptTmp = this.clone()
  if (truncate === null) {
    truncate = true;
  }

  if ('undefined' === typeof maxEl) {
    maxEl = 20;
  }

  var type = scriptTmp.getOutType()
  var s = '';
  for (var i = 0, l = scriptTmp.chunks.length; i < l; i++) {
    var chunk = scriptTmp.chunks[i];

    if (i > 0) {
      s += ' ';
    }

    if (Array.isArray(chunk)) {
      if (truncate === true) {
        var maxLen = chunk.length
      } else {
        var maxLen = 100
      }
      s += new Buffer(chunk.splice(0, maxLen)).toString('hex');

      if (truncate) {
        s += '...'
      }
    } else {
      if (type === 'multisig') {
        switch (chunk) {
          case 80:
            s += 'OP_0'
            break
          case 81:
            s += 'OP_1'
            break
          default:
            s += Opcode.reverseMap[chunk]
        }
      } else {
        s += Opcode.reverseMap[chunk];
      }
    }

    if (maxEl && i > maxEl) {
      s += ' ...';
      break;
    }
  }
  return s;
};
